### PR TITLE
Merge master into LOG-8778

### PR DESCRIPTION
**Pull request by jyeung-r7**

### CODE REVIEW

File: src/le.py [hunk 1]
```diff
@@ -1729,13 +1779,81 @@ def cmd_ls_ips():
...
+    response = api_v2_request('GET', 'structures', {}, True)
+    structures = response['structures']
+
+    for structure in sorted(structures, key=lambda x: x['name']):
+        if CONFIG.uuid:
+            print (utils.c_id(structure['id']))
+        print (structure['name'])
+
+    if len(structures) == 0:
+        print >> sys.stderr, 'No structures defined'
```
Comment (jsouthern-r7) [inline src/le.py]: This isn't Python 3 compatible. Consider replacing with...

`sys.stderr.write('No structures defined')`

...or...

`print('No structures defined', file=sys.stderr)`

File: src/le.py [hunk 2]
```diff
@@ -1729,13 +1779,81 @@ def cmd_ls_ips():
...
+    elif len(structures) == 1:
+        print >> sys.stderr, '1 structure'
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

`print('1 structure', file=sys.stderr)`

File: src/le.py [hunk 3]
```diff
@@ -1729,13 +1779,81 @@ def cmd_ls_ips():
...
+    else:
+        print >> sys.stderr, '%s structures' % len(structures)
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

`print('%s structures' % len(structures), file=sys.stderr)`

File: src/le.py [hunk 4]
```diff
@@ -1729,13 +1779,81 @@ def cmd_ls_ips():
...
+
+    response = api_v2_request('GET', 'structures/%s/patterns'%structure_id, {}, True)
+
+    patterns = response['patterns']
+    for pattern in sorted(patterns, cmp=utils.cmp_patterns):
+        if CONFIG.uuid:
+            print (utils.c_id(pattern['id'])),
+        print ('%2d %s' % (pattern['priority'], pattern['pattern']))
+    if len(patterns) == 0:
+        print >> sys.stderr, 'No patterns in structure `%s\'' % structure_name
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

`print('No patterns in structure `%s\'' %structure_name, file=sys.stderr)`

File: src/le.py [hunk 5]
```diff
@@ -1729,13 +1779,81 @@ def cmd_ls_ips():
...
+    elif len(patterns) == 1:
+        print >> sys.stderr, '1 pattern'
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

`print('1 pattern', file=sys.stderr)`

File: src/le.py [hunk 6]
```diff
@@ -1729,13 +1779,81 @@ def cmd_ls_ips():
...
+    else:
+        print >> sys.stderr, '%s patterns' % len(patterns)
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

print('%s patterns' % len(patterns), file=sys.stderr)

File: src/le.py [hunk 7]
```diff
@@ -1752,11 +1870,178 @@ def cmd_ls(args):
...
+    matches = [s['id'] for s in structures if structure_name.lower() == s['id'].lower() or structure_name == s['name']]
+    if len(matches) == 0:
+        # Add the structure
+        response = api_v2_request('POST', 'structures', {
+            'name': structure_name,
+            'shortcut': structure_name,
+            'description': '',
+            'aux': {},
+            }, True)
+        print >> sys.stderr, 'Added structure `%s\'' % structure_name
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

print('Added structure `%s\'' % structure_name,  file=sys.stderr)

File: src/le.py [hunk 8]
```diff
@@ -1752,11 +1870,178 @@ def cmd_ls(args):
...
+        structure_id = response['structure']['id']
+    else:
+        structure_id = matches[0]
+        if not pattern:
+            print >> sys.stderr, 'Structure `%s\' already exists' % structure_name
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

```
print('Structure `%s\' already exists' % structure_name, file=sys.stderr)
111

File: src/le.py [hunk 9]
```diff
@@ -1752,11 +1870,178 @@ def cmd_ls(args):
...
+
+    # Add pattern (if specified)
+    # TODO - check that the pattern does not exist already
+    if pattern:
+        response = api_v2_request('POST', 'structures/%s/patterns'%structure_id, {
+            'priority': priority,
+            'pattern': pattern,
+            }, True)
+        print >> sys.stderr, 'Added pattern `%s\'' % pattern
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

```
print('Added pattern `%s\'' % pattern, file=sys.stderr)
```

File: src/le.py [hunk 10]
```diff
@@ -1752,11 +1870,178 @@ def cmd_ls(args):
...
+    response = api_v2_request('GET', 'structures/%s/patterns'%structure_id, {}, True)
+    patterns = response['patterns']
+
+    # Go though patterns one by one, identify pattern IDs
+    # Use a different call for pattern=*
+    matches = [[p['id'], p['pattern']] for p in patterns \
+            if rm_all or p['id'] == pattern or p['pattern'].startswith(pattern)]
+    if not matches:
+        if rm_all:
+            print >> sys.stderr, 'No pattern to be removed; structure is empty'
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

```
print('No pattern to be removed; structure is empty', file=sys.stderr)
```

File: src/le.py [hunk 11]
```diff
@@ -1752,11 +1870,178 @@ def cmd_ls(args):
...
+        else:
+            utils.error('No pattern matching `%s\' found', pattern)
+    if len(matches) > 1 and not rm_all:
+        utils.error('Pattern `%s\' is not specific enough; multiple matches', pattern)
+
+    # Do the physical deletion
+    for pattern_id, pattern_p in matches:
+        response = api_v2_request('DELETE', 'structures/%s/patterns/%s'%(structure_id, pattern_id), {}, True)
+        print >> sys.stderr, 'Removed pattern `%s\'' % pattern_p
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

```
print('Removed pattern `%s\'' % pattern_p, file=sys.stderr)
```

File: src/le.py [hunk 12]
```diff
@@ -1752,11 +1870,178 @@ def cmd_ls(args):
...
+        utils.error('Structure `%s\' does not exist', structure_name)
+    if len(structure) > 1:
+        utils.error('Multiple matches for `%s\'', structure_name)
+    structure_id = structure[0][1]
+
+    # Now we know structure exists and we know its ID
+
+    if not patterns: # Removing structure
+        response = api_v2_request('DELETE', 'structures/%s'%structure_id, {}, True)
+        print >> sys.stderr, 'Removed structure `%s\'' % structure_name
```
Comment (jsouthern-r7) [inline src/le.py]: Consider replacing with...

```
print('Removed structure `%s\'' % structure_name, file=sys.stderr)
```

**Comment by ellynch-r7** [comment]
We should probably add in some tests for this.

Ideally, we want to start using pytest (or similar) so that the tests can be run easily on any platform.

**Pull request merged**